overlord/hookstate/hookmgr.go

```diff
@@ -296,16 +323,36 @@ func runHookAndWait(snapName string, revision snap.Revision, hookName, hookConte
 		close(hookCompleted)
 	}()
 
+	var abortOrTimeoutError error
 	select {
-	// Hook completed; it may or may not have been successful.
 	case <-hookCompleted:
+		// Hook completed; it may or may not have been successful.
 		return buffer.Bytes(), hookError
-
-	// Hook was aborted.
 	case <-tomb.Dying():
-		if err := command.Process.Kill(); err != nil {
-			return nil, fmt.Errorf("cannot abort hook %q: %s", hookName, err)
-		}
-		return nil, fmt.Errorf("hook %q aborted", hookName)
+		// Hook was aborted, process will get killed below
+		abortOrTimeoutError = fmt.Errorf("hook %q aborted", hookName)
+	case <-killTimerCh:
+		// Max timeout reached, process will get killed below
+		abortOrTimeoutError = fmt.Errorf("exceeded maximum runtime of %s", timeout)
+	}
+
+	// select above exited which means that aborted or killTimeout
+	// was reached. Kill the command and wait for command.Wait()
+	// to clean it up (but limit the wait with the cmdWaitTimer)
+	if err := killemAll(command); err != nil {
+		return nil, fmt.Errorf("cannot abort hook %q: %s", hookName, err)
 	}
+	select {
+	case <-time.After(cmdWaitTimeout):
+		// cmdWaitTimeout was reached, i.e. command.Wait() did not
+		// finish in a reasonable amount of time, we can not use
+		// buffer in this case so return without it.
+		return nil, fmt.Errorf("%v, but did not stop", abortOrTimeoutError)
+	case <-hookCompleted:
+		// cmd.Wait came back from waiting the killed process
+		break
+	}
+	fmt.Fprintf(buffer, "\n%s", abortOrTimeoutError)
```

Comment:
maybe "\n<%s>" ?  my worry is that it looks too much like from the script, vs from snapd ?